cdap-tms/src/main/java/co/cask/cdap/messaging/store/hbase/HBaseMessageTable.java

```diff
@@ -0,0 +1,234 @@
+ */
+public class HBaseMessageTable implements MessageTable {
+  private static final byte[] PAYLOAD_COL = Bytes.toBytes('p');
+  private static final byte[] TX_COL = Bytes.toBytes('t');
+  private final Configuration hConf;
+  private final HBaseTableUtil tableUtil;
+  private final TableId tableId;
+  private final byte[] columnFamily;
+  private volatile long writeTimestamp;
+  private volatile short seqId;
+  private volatile HTable hTable;
+
+  public HBaseMessageTable(Configuration hConf, HBaseTableUtil tableUtil, TableId tableId, byte[] columnFamily) {
+    this.hConf = hConf;
+    this.tableUtil = tableUtil;
+    this.tableId = tableId;
+    this.columnFamily = Arrays.copyOf(columnFamily, columnFamily.length);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, long startTime, int limit, @Nullable Transaction transaction)
+    throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(startTime));
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(startRow)
+      .setStopRow(Bytes.stopKeyForPrefix(topic))
+      .build();
+    ResultScanner scanner = getHTable().getScanner(scan);
+    return new HBaseCloseableIterator(scanner, limit, columnFamily, transaction);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, MessageId messageId, boolean inclusive,
+                                        int limit, @Nullable Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(messageId.getPublishTimestamp()),
+                                Bytes.toBytes(messageId.getSequenceId()));
+    if (!inclusive) {
+      startRow = Bytes.incrementBytes(startRow, 1);
+    }
+    Scan scan = tableUtil.buildScan()
+      .setStartRow(startRow)
+      .setStopRow(Bytes.stopKeyForPrefix(topic))
+      .build();
+    ResultScanner scanner = getHTable().getScanner(scan);
+    return new HBaseCloseableIterator(scanner, limit, columnFamily, transaction);
+  }
+
+  @Override
+  public void store(Iterator<Entry> entries) throws IOException {
+    long writeTs = System.currentTimeMillis();
+    if (writeTs != writeTimestamp) {
+      seqId = 0;
+    }
+    writeTimestamp = writeTs;
+    List<Put> batchPuts = new ArrayList<>();
+    while (entries.hasNext()) {
+      Entry entry = entries.next();
+      byte[] topic = MessagingUtils.toBytes(entry.getTopicId());
+      byte[] tableKey = Bytes.add(topic, Bytes.toBytes(writeTimestamp), Bytes.toBytes(seqId++));
+      PutBuilder putBuilder = tableUtil.buildPut(tableKey);
+      if (entry.isPayloadReference()) {
+        byte[] txPtrBytes = Bytes.toBytes(entry.getTransactionWritePointer());
+        putBuilder.add(columnFamily, TX_COL, txPtrBytes);
+      } else {
+        putBuilder.add(columnFamily, PAYLOAD_COL, entry.getPayload());
+        if (entry.isTransactional()) {
```

Comment:
Seems like this can be pulled out as

```java
if (entry.isTransactional()) {
  putBuilder.add(columnFamily, TX_COL, Bytes.toBytes(entry.getTransactionWritePointer()));
}
if (!entry.isPayloadReference()) {
   putBuilder.add(columnFamily, PAYLOAD_COL, entry.getPayload());
}